### PR TITLE
Refactor Lavaland river spawner.

### PR DESCRIPTION
**Pull request by warriorstar-orion**
## What Does This PR Do

This PR performs a minimally invasive refactor of the "river" generator used for Lavaland for lava/plasma/chasms. This is the first part of reworking the generator to be useful for other kinds of generation, because at its core it is a very useful implementation for meandering paths of turfs, and is easily generalizable with not too many changes.

I split up the enormous signature for `/proc/spawn_rivers()` in a way I felt made sense. The very basic configuration is the constructor, which includes the target z-level and spread behavior and falloff. The actual `generate()` call takes the number of desired waypoint nodes, and the X, Y bounding box for the river. This makes it easy to reuse the same generator multiple times if you want it to be used in a different location.

<details><summary>More information about the "minor deviation" in behavior mentioned below</summary>
At some point when this code was still in TG, a change was made to add riverbanks to the spawner (what I call shorelines in this PR, because I was unaware of the original naming). This introduced the logged_turf_type you can see in the original code, which is set to the first mineral wall's turf_tile when checking adjacent turfs to see if it should flip them into the river turf. a mineral wall's turf_tile is effectively its corresponding floor tile: what it should turn into when dug.

image

However, this doesn't work as it should in para. I thiiiiiiiiiiiiink it has to do with the fact that our typepath hierarchy is different. Whereas its original job is to find the first nearby acceptable floor tile to use as the riverbank, what happens with our code is that it's able to double back on tiles that were already flipped into river tiles, and call Spread on them again as you can see in this log output.

image

And because RANGE_TURFS uses locate which returns turfs in a deterministic order, this causes an extremely subtle bias, where in certain tile arrangements, Spread() would double-back on itself before continuing. My hunch is that the change in behavior to what I have now (where it just checks to see if its an acceptable turf to turn into a shoreline) is impercetible if not completely indistinguisable. but if that is true, then we're just waiting time calling Spread on tiles we already called Spread on.
</details>

## Why It's Good For The Game

The current procs are horribly coupled to the generator's current, specific usage.

1. `/turf/proc/Spread()` is far too concerned with its subtypes, using checks like `ismineralturf()` and performing casts to `/turf/simulated/mineral`. On top of that, attempting to call this on basically any other subtype of `/turf` than what it's currently used for would almost certainly fail.
2. It relies on `/turf/simulated/mineral/var/turf_type` to choose what I ended up calling the "shoreline" of rivers. Not only is this var buried in a subtype, but it then served dual functions as both the "shoreline" and its original purpose as the replacement turf once a mineral wall gets mined, which you may not want to be the same for any number of reasons.
3. `/turf/proc/Spread()` is also too concerned with spreading types conceived as dangerous floor tiles, such as spawning bridges under certain circumstances, which wouldn't make sense if you were trying to spawn and spread, say, walls, or even floors that aren't dangerous (I haven't changed this behavior here but it is now much easier to toggle in the future).
4. `/turf/simulated/mineral` had its own implementation of `Spread()` with a _completely different proc signature_, so it couldn't even be used in `/proc/spread_rivers()` without more special-casing, because, turns out, calling `ChangeTurf` on the integer 25 doesn't work.

## Images of changes

With only one incredibly minor deviation from the original implementation, the resultant rivers should look exactly the same. Here are some screenshots regardless.

2024_02_16__02_18_47__ C__Users_warriorstar_iCloudDrive_Screenshots_2024_02_16_2024_02_16__01_24_22__P
2024_02_16__02_19_02__ C__Users_warriorstar_iCloudDrive_Screenshots_2024_02_16_2024_02_16__01_24_29__P
2024_02_16__02_31_51__ C__Users_warriorstar_iCloudDrive_Screenshots_2024_02_16_2024_02_16__02_28_33__P
2024_02_16__02_32_02__ C__Users_warriorstar_iCloudDrive_Screenshots_2024_02_16_2024_02_16__02_29_07__P
2024_02_16__02_32_57__ C__Users_warriorstar_iCloudDrive_Screenshots_2024_02_16_2024_02_16__02_32_30__P
2024_02_16__02_37_59__ C__Users_warriorstar_iCloudDrive_Screenshots_2024_02_16_2024_02_16__02_37_44__P

## Testing

Over a hundred fucking lavaland generations, not even exaggerating.

## Changelog

NPFC

### CODE REVIEW

File: code/game/turfs/simulated/river.dm [hunk 1]
```diff
@@ -1,17 +1,52 @@
...
 
-#define RANDOM_LOWER_X 50
-#define RANDOM_LOWER_Y 50
+#define RIVER_MIN_X 50
+#define RIVER_MIN_Y 50
 
-/proc/spawn_rivers(target_z, nodes = 4, turf_type = /turf/simulated/floor/lava/mapping_lava, whitelist_area = /area/lavaland/surface/outdoors, min_x = RANDOM_LOWER_X, min_y = RANDOM_LOWER_Y, max_x = RANDOM_UPPER_X, max_y = RANDOM_UPPER_Y, prob = 25, prob_loss = 11)
+/obj/effect/landmark/river_waypoint
+	name = "river waypoint"
+	var/connected = FALSE
```
Comment (lewcc) [inline code/game/turfs/simulated/river.dm:9]: Can you document this var?

File: code/modules/lavaland/lavaland_theme.dm [hunk 2]
```diff
@@ -25,32 +25,24 @@
...
 	name = "plasma"
 	primary_turf_type = /turf/simulated/floor/lava/lava_land_surface/plasma
 	planet_icon_state = "planet_plasma"
 
 /datum/lavaland_theme/plasma/setup()
-	spawn_rivers(level_name_to_num(MINING), nodes = 2)
-	spawn_rivers(level_name_to_num(MINING), nodes = 2)
+	var/datum/river_spawner/spawner = new(level_name_to_num(MINING))
+	spawner.generate(nodes = 2)
+	spawner.generate(nodes = 2) // twice
```
Comment (lewcc) [inline code/modules/lavaland/lavaland_theme.dm:39]: Why twice? If this is just a relic of the past, that's fine, I'm just curious if you know the motivation behind it

**Review by lewcc — COMMENTED** [review]
Looks pretty good, just some minor nits

File: code/game/turfs/simulated/river.dm (hunk 1 above)
Comment (warriorstar-orion) [reply code/game/turfs/simulated/river.dm:10]: Done

**Review by S34NW — APPROVED** [review]

**Pull request merged**